File: Week 1 C/credit.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

bool is_valid_credit_card(const char *card_number);

int main()
{
    char card_number[20];
    printf("Enter credit card number: ");
    scanf("%s", card_number);

    if (is_valid_credit_card(card_number))
    {
        if (strncmp(card_number, "34", 2) == 0 || strncmp(card_number, "37", 2) == 0)
        {
            printf("American Express\n");
        }
        else if (strncmp(card_number, "51", 2) == 0 || strncmp(card_number, "52", 2) == 0 ||
                 strncmp(card_number, "53", 2) == 0 || strncmp(card_number, "54", 2) == 0 ||
                 strncmp(card_number, "55", 2) == 0)
        {
            printf("Mastercard\n");
        }
        else if (strncmp(card_number, "4", 1) == 0)
        {
            printf("Visa\n");
        }
        else
        {
            printf("Invalid\n");
        }
    }
    else
    {
        printf("Invalid\n");
    }

    return 0;
}

bool is_valid_credit_card(const char *card_number)
{
    int len = strlen(card_number);
    int sum = 0;
    bool double_digit = false;

    for (int i = len - 1; i >= 0; i--)
    {
        int digit = card_number[i] - '0';
        if (double_digit)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }
        sum += digit;
        double_digit = !double_digit;
    }

    return (sum % 10 == 0);
}

File: Week 6 Python/sentimental-credit/credit.py
# This program checks the validity of credit card numbers

# Returns True if the credit card number is valid, False otherwise
def is_valid_credit_card(card_number):
    card_number = card_number[::-1]
    sum = 0
    double_digit = False
    for i in range(len(card_number)):
        digit = int(card_number[i])
        if double_digit:
            digit *= 2
            if digit > 9:
                digit -= 9
        sum += digit
        double_digit = not double_digit
    return sum % 10 == 0


if __name__ == "__main__":
    # Get credit card number from user input
    card_number = input("Enter credit card number: ").strip()

    # Determine the card's issuer
    if is_valid_credit_card(card_number):
        if card_number.startswith("34") or card_number.startswith("37"):
            print("AMEX")
        elif card_number.startswith(("51", "52", "53", "54", "55")):
            print("MASTERCARD")
        elif card_number.startswith("4") and (len(card_number) == 13 or len(card_number) == 16):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")